apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java

```diff
@@ -121,101 +119,31 @@ public static void main(String[] args) {
-        new DataSetMonitorListener() {
-          @Override
-          public void dataAvailable(RifFilesEvent rifFilesEvent) {
-            Timer.Context timerDataSet =
-                appMetrics
-                    .timer(
-                        MetricRegistry.name(
-                            PipelineApplication.class.getSimpleName(), "dataSet", "processed"))
-                    .time();
-
-            Consumer<Throwable> errorHandler =
-                error -> {
-                  /*
-                   * This will be called on the same thread used to run each
-                   * RifLoader task (probably a background one). This is not
-                   * the right place to do any error _recovery_ (that'd have
-                   * to be inside RifLoader itself), but it is likely the
-                   * right place to decide when/if a failure is "bad enough"
-                   * that the rest of processing should be stopped. Right now
-                   * we stop that way for _any_ failure, but we probably want
-                   * to be more discriminating than that.
-                   */
-                  errorOccurred(error);
-                };
-
-            Consumer<RifRecordLoadResult> resultHandler =
-                result -> {
-                  /*
-                   * Don't really *need* to do anything here. The RifLoader
-                   * already records metrics for each data set.
-                   */
-                };
-
-            /*
-             * Each ETL stage produces a stream that will be handed off to
-             * and processed by the next stage.
-             */
-            for (RifFileEvent rifFileEvent : rifFilesEvent.getFileEvents()) {
-              Slf4jReporter dataSetFileMetricsReporter =
-                  Slf4jReporter.forRegistry(rifFileEvent.getEventMetrics())
-                      .outputTo(LOGGER)
-                      .build();
-              dataSetFileMetricsReporter.start(2, TimeUnit.MINUTES);
-
-              RifFileRecords rifFileRecords = rifProcessor.produceRecords(rifFileEvent);
-              rifLoader.process(rifFileRecords, errorHandler, resultHandler);
-
-              dataSetFileMetricsReporter.stop();
-              dataSetFileMetricsReporter.report();
-            }
-            timerDataSet.stop();
-          }
-
-          /**
-           * @see
-           *     gov.cms.bfd.pipeline.ccw.rif.extract.s3.DataSetMonitorListener#errorOccurred(java.lang.Throwable)
-           */
-          @Override
-          public void errorOccurred(Throwable error) {
-            handleUncaughtException(error);
-          }
-
-          /** Called when no RIF files are available to process. */
-          @Override
-          public void noDataAvailable() {
-            rifLoader.doIdleTask();
-            DataSetMonitorListener.super.noDataAvailable();
-          }
-        };
+    pipelineManager.registerJob(new DatabaseSchemaUpdateJob(appConfig.getDatabaseOptions()));
```

Comment:
Just thinking out loud, but this is the sort of thing that the Java Service Provider Interfaces are designed for.  That would allow new jobs to be discovered and picked up by just dropping them on the classpath for the `PipelineApplication`.

The code here definitely a more explicit approach, and I can see the value of that though but it feels like a "plugin" architecture could make sense here.

Reply:
I think heading down the SPI/plugin route would _also_ require a dependency injection provider, e.g. Spring Beans or CDI or Guice, right? I dunno'; I should be in bed, so perhaps I'm wrong.

If so, though, I've very intentionally avoided that here. I used to be a big Guice/Dagger stan but these days find that they add way more complexity than they remove.

Even aside from that, given the small number of jobs we have (and are likely to have for the next couple years), I think I prefer how easy it is to figure out from what's going on just by looking at `PipelineApplication`. That would definitely need to be re-evaluated if the number of jobs grown past some point, e.g. towards 10.

Reply:
No Java SPI doesn't require CDI.  It is in the base spec.

Reply:
I'm going to defer this, at least for now. At the moment, I'm more in favor of less magic (heh, wordplay). It's a good suggestion, though, and one worth considering in the future, particularly if we end up with large numbers of jobs.